uname/uname.1
.Dd December 2008
.Dt UNAME 1

.Sh NAME

.Nm uname
.Nd return system name

.Sh SYNOPSIS

.Nm uname
.Ar [-snrvma]

.Sh DESCRIPTION

    By default, the uname utility shall write the operating system name to
standard output. When options are specified, symbols representing one or more
system characteristics shall be written to the standard output. The format
and contents of the symbols are implementation-defined. On systems conforming
to the System Interfaces volume of IEEE Std 1003.1-2001, the symbols written
shall be those supported by the uname() function as defined in the System
Interfaces volume of IEEE Std 1003.1-2001.

.Sh OPTIONS

    The uname utility shall conform to the Base Definitions volume of IEEE
Std 1003.1-2001, Section 12.2, Utility Syntax Guidelines.

    The following options shall be supported:

    -a
        Behave as though all of the options -mnrsv were specified.
    -m
        Write the name of the hardware type on which the system is running to
standard output.
    -n
        Write the name of this node within an implementation-defined
communications network.
    -r
        Write the current release level of the operating system
implementation.
    -s
        Write the name of the implementation of the operating system.
    -v
        Write the current version level of this release of the operating
system implementation.

    If no options are specified, the uname utility shall write the operating
system name, as if the -s option had been specified.

.Sh OPERANDS

    None.

.Sh STDIN

    Not used.

.Sh INPUT FILES

    None.

.Sh ENVIRONMENT VARIABLES

    The following environment variables shall affect the execution of uname:

    LANG
        Provide a default value for the internationalization variables that
are unset or null. (See the Base Definitions volume of IEEE Std 1003.1-2001,
Section 8.2, Internationalization Variables for the precedence of
internationalization variables used to determine the values of locale
categories.)
    LC_ALL
        If set to a non-empty string value, override the values of all the
other internationalization variables.
    LC_CTYPE
        Determine the locale for the interpretation of sequences of bytes of
text data as characters (for example, single-byte as opposed to multi-byte
characters in arguments).
    LC_MESSAGES
        Determine the locale that should be used to affect the format and
contents of diagnostic messages written to standard error.
    NLSPATH
        [XSI] [Option Start] Determine the location of message catalogs for
the processing of LC_MESSAGES . [Option End]

.Sh ASYNCHRONOUS EVENTS

    Default.

.Sh STDOUT

    By default, the output shall be a single line of the following form:

    "%s\n", <sysname>

    If the -a option is specified, the output shall be a single line of the
following form:

    "%s %s %s %s %s\n", <sysname>, <nodename>, <release>,
        <version>, <machine>

    Additional implementation-defined symbols may be written; all such
symbols shall be written at the end of the line of output before the
<newline>.

    If options are specified to select different combinations of the symbols,
only those symbols shall be written, in the order shown above for the -a
option. If a symbol is not selected for writing, its corresponding trailing
<blank>s also shall not be written.

.Sh STDERR

    The standard error shall be used only for diagnostic messages.

.Sh OUTPUT FILES

    None.

.Sh EXTENDED DESCRIPTION

    None.

.Sh EXIT STATUS

    The following exit values shall be returned:

     0
        The requested information was successfully written.
    >0
        An error occurred.

.Sh CONSEQUENCES OF ERRORS

    Default.

The following sections are informative.
.Sh APPLICATION USAGE

    Note that any of the symbols could include embedded <space>s, which may
affect parsing algorithms if multiple options are selected for output.

    The node name is typically a name that the system uses to identify itself
for inter-system communication addressing.

.Sh EXAMPLES

    The following command:

    uname -sr

    writes the operating system name and release level, separated by one or
more <blank>s.

.Sh RATIONALE

    It was suggested that this utility cannot be used portably since the
format of the symbols is implementation-defined. The POSIX.1 working group
could not achieve consensus on defining these formats in the underlying
uname() function, and there was no expectation that this volume of IEEE Std
1003.1-2001 would be any more successful. Some applications may still find
this historical utility of value. For example, the symbols could be used for
system log entries or for comparison with operator or user input.

.Sh FUTURE DIRECTIONS

    None.

.Sh SEE ALSO

    The System Interfaces volume of IEEE Std 1003.1-2001, uname()



NEWS

Version 5 - unreleased

	- New utilities: cmp, tput, tabs, stty, uname, uniq

	- Bug fixes.

Version 4 - December 9, 2006

	- import codebase into git repository on git.kernel.org
	- fix compiler warnings on newer compilers
	- add x86-64 support
	- add WIP utility: pax
	- add utilities: env, id, strings, time
	- add simple "it works" tests
	- create per-util status files
	- drop hand-rolled 'compress', import BSD 'compress' instead

Version 3 - February 7, 2005

	- Add copyright/license info to each source file.
	- Make sure bug report address points to SourceForge.

	- New utilities: grep, pathchk, pwd, split, tty.
	- New work-in-progress (WIP) utility: mv.

	- Fix many 'compress' bugs.
	- Security fixes.

	- (internal) Move compress engine into lib.

Version 2 - January 26, 2005

	- Update README with project goals/manifesto,
	  and some URLs.

	- New utilities: compress, comm, date, what, wc, renice,
	  ipcrm, kill, rm.

	- Bug fixes.

Version 1 - August 21, 2004

	- Initial release.



uname/uname.cc

/*
 * Copyright 2008-2009 Red Hat, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */


#ifndef HAVE_CONFIG_H
#error missing autoconf-generated config.h.
#endif
#include "posixutils-config.h"

#include <sys/types.h>
#include <sys/utsname.h>
#include <stdlib.h>
#include <string.h>
#include <argp.h>
#include <libpu.h>


static const char doc[] =
N_("uname - return system name");

static struct argp_option options[] = {
	{ "all", 'a', NULL, 0,
	  N_("Behave as though all of the options -mnrsv were specified.") },
	{ "machine", 'm', NULL, 0,
	  N_("Write the name of the hardware type on which the system is running to standard output.") },
	{ "nodename", 'n', NULL, 0,
	  N_("Write the name of this node within an implementation-defined communications network.") },
	{ "kernel-release", 'r', NULL, 0,
	  N_("Write the current release level of the operating system implementation.") },
	{ "kernel-name", 's', NULL, 0,
	  N_("Write the name of the implementation of the operating system.") },
	{ "kernel-version", 'v', NULL, 0,
	  N_("Write the current version level of this release of the operating system implementation.") },
	{ }
};

static error_t parse_opt (int key, char *arg, struct argp_state *state);
static const struct argp argp = { options, parse_opt, NULL, doc };

enum parse_options_bits {
	OPT_MACHINE		= (1 << 1),
	OPT_NODE		= (1 << 2),
	OPT_KREL		= (1 << 3),
	OPT_KNAME		= (1 << 4),
	OPT_KVER		= (1 << 5),

	OPT_ALL			= (OPT_MACHINE | OPT_NODE | OPT_KREL |
				   OPT_KNAME | OPT_KVER),
};

static int opt_mask;

static error_t parse_opt (int key, char *arg, struct argp_state *state)
{
	switch (key) {
	case 'a': opt_mask |= OPT_ALL; break;
	case 'm': opt_mask |= OPT_MACHINE; break;
	case 'n': opt_mask |= OPT_NODE; break;
	case 'r': opt_mask |= OPT_KREL; break;
	case 's': opt_mask |= OPT_KNAME; break;
	case 'v': opt_mask |= OPT_KVER; break;

	default:
		return ARGP_ERR_UNKNOWN;
	}

	return 0;
}

int main (int argc, char *argv[])
{
	char outstr[1024];
	struct utsname uts;

	pu_init();

	error_t argp_rc = argp_parse(&argp, argc, argv, 0, NULL, NULL);
	if (argp_rc) {
		fprintf(stderr, _("%s: argp_parse failed: %s\n"),
			argv[0], strerror(argp_rc));
		return EXIT_FAILURE;
	}

	if (!opt_mask)
		opt_mask = OPT_KNAME;

	if (uname(&uts) < 0) {
		perror("uname");
		return 1;
	}

	outstr[0] = 0;

	if (opt_mask & OPT_MACHINE) {
		strcat(outstr, uts.machine);
		strcat(outstr, " ");
	}
	if (opt_mask & OPT_NODE) {
		strcat(outstr, uts.nodename);
		strcat(outstr, " ");
	}
	if (opt_mask & OPT_KREL) {
		strcat(outstr, uts.release);
		strcat(outstr, " ");
	}
	if (opt_mask & OPT_KNAME) {
		strcat(outstr, uts.sysname);
		strcat(outstr, " ");
	}
	if (opt_mask & OPT_KVER) {
		strcat(outstr, uts.version);
		strcat(outstr, " ");
	}
	if (outstr[0])
		outstr[strlen(outstr) - 1] = '\n';

	printf("%s", outstr);

	return 0;
}

